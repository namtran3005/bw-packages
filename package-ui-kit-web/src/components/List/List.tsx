import React from 'react';
import styled from 'styled-components';
import { Spacing, Colors, ColorsType } from '@bitwala-cryptobank-squad/package-theme';
import { List as MuiList } from '@material-ui/core';
import { ListProps as MuiListProps } from '@material-ui/core/List';
import { SpacingType, getPxValue } from '../Content/Content.styled';
import {
  MediaStyledProperty,
  mediaStyledProperties,
} from '../../utils/mediaStyledProperty';

import {
  Root as ListItem,
  highlightWidth,
  doubleHighlightWidth,
} from './ListItem';
interface RootProps extends MuiListProps {
  innerRef?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  variant?:
    | 'sidepanel'
    | 'normal'
    | 'highlighted'
    | 'doubleHighlighted'
    | 'info'
    | 'recipient'
    | 'wealth'
    | 'transaction'
    | 'default';

  foreground?: ColorsType;
  hoverForeground?: ColorsType;
  selectedForeground?: ColorsType;
  highlightForeground?: ColorsType;
  topSpacing?: MediaStyledProperty<SpacingType>;
  bottomSpacing?: MediaStyledProperty<SpacingType>;
  spacing?: MediaStyledProperty<SpacingType>;
  component?: React.ElementType;
}

const defaultForeground: ColorsType = 'primaryBlack';
const defaultHoverForeground: ColorsType = 'primaryBlack';
const defaultSelectedForeground: ColorsType = 'primaryBlack';

const RootComponent: React.FunctionComponent<RootProps> = ({
  /* eslint-disable @typescript-eslint/no-unused-vars */
  variant,
  foreground,
  hoverForeground,
  selectedForeground,
  highlightForeground,
  topSpacing,
  bottomSpacing,
  spacing,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ...cleanProps
}) => <MuiList {...cleanProps} />;

const Root = styled(RootComponent)`
  &&& {
    position: relative;
    color: ${({ foreground }) => Colors.get(foreground || defaultForeground)};
    padding-top: ${Spacing.small}px;

    ${({ topSpacing, bottomSpacing, spacing = 8 }) =>
      mediaStyledProperties({
        'padding-top': {
          value: typeof topSpacing !== 'undefined' ? topSpacing : spacing,
          transform: getPxValue,
        },
        'padding-bottom': {
          value: typeof bottomSpacing !== 'undefined' ? bottomSpacing : spacing,
          transform: getPxValue,
        },
      })};

    a {
      text-decoration: none;
    }

    ${ListItem} {
      color: ${({ variant, foreground }) =>
        ['highlighted', 'doubleHighlighted', 'sidepanel'].indexOf(
          variant || 'normal'
        ) === -1
          ? 'inherit'
          : Colors.get(foreground || defaultForeground)};
      flex-direction: ${({ variant }) => variant === 'info' && 'column'};
      align-items: ${({ variant }) => variant === 'info' && 'flex-start'};

      &:after,
      &:before {
        background: ${({ highlightForeground }) =>
          highlightForeground
            ? Colors.get(highlightForeground)
            : 'currentColor'};
      }

      &:after {
        opacity: ${({ variant }) =>
          ['highlighted', 'doubleHighlighted'].indexOf(variant as string) !== -1
            ? 1
            : 0};
      }

      &:before {
        opacity: ${({ variant }) =>
          variant === 'doubleHighlighted' ? 0.2 : 0};
      }

      &:hover {
        box-shadow: 0 0 ${Spacing.unit * 4}px
          ${Colors.getWithAlpha('primaryBlack', 0.1)};

        color: ${({ hoverForeground }) =>
          Colors.get(hoverForeground || defaultHoverForeground)};

        &:after {
          width: ${highlightWidth}px;
        }

        &:before {
          width: ${doubleHighlightWidth}px;
        }
      }

      &.selected {
        color: ${({ selectedForeground }) =>
          Colors.get(selectedForeground || defaultSelectedForeground)};
      }
    }
  }
`;

Root.displayName = 'List';
RootComponent.displayName = 'ListComponent';

const List: React.FC<RootProps> = ({ children, ...props }) => {
  const childrenWithProps = React.Children.map(children, (child) => {
    return child
      ? React.cloneElement(child as React.ReactElement<any>, {
          variant: props.variant,
        })
      : null;
  });

  return <Root {...props}>{childrenWithProps}</Root>;
};

export { List, RootProps as ListProps };
