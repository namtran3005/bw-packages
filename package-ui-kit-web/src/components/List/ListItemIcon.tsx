import * as React from 'react';
import styled from 'styled-components';
import { Spacing } from '@bitwala-cryptobank-squad/package-theme';

interface RootProps {
  key?: number | string;
}

const RootComponent: React.FunctionComponent<RootProps> = ({
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  key,
  ...cleanProps
}) => <div {...cleanProps} />;

const Root = styled(RootComponent)`
  color: inherit;
  margin-right: ${Spacing.unit * 2}px;

  & > * {
    vertical-align: middle;
  }
`;

Root.displayName = 'ListItemIcon';

export { Root, RootProps, RootComponent };

const ListItemIcon: React.FC<RootProps> = Root;

export { ListItemIcon, RootProps as ListItemIconProps };
