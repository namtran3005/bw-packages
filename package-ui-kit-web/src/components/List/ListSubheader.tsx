import * as React from 'react';
import styled from 'styled-components';

import { Spacing } from '@bitwala-cryptobank-squad/package-theme';
import { Typography, TypographyProps } from '../Typography';
import { mediaStyledProperties } from '../../utils/mediaStyledProperty';
import { RootSpacingProps, getPxValue } from '../Content/Content.styled';

interface RootProps extends TypographyProps, RootSpacingProps {
  foreground?: TypographyProps['foreground'];
  fontWeight?: TypographyProps['fontWeight'];
  typographyProps?: TypographyProps;
}

const RootComponent: React.FC<RootProps> = ({
  foreground,
  fontWeight,
  typographyProps,
  children,
  className,
}) => (
  <div className={className}>
    <Typography.Small
      foreground={foreground || 'subtitleText'}
      fontWeight={fontWeight || 'bold'}
      {...typographyProps}
    >
      {children}
    </Typography.Small>
  </div>
);

const Root = styled(RootComponent)`
  &&& {
    text-transform: uppercase;
    ${({ topSpacing, bottomSpacing, leftSpacing, rightSpacing, spacing }) =>
      mediaStyledProperties({
        'padding-top': {
          value:
            typeof topSpacing !== 'undefined'
              ? topSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.unit,
          transform: getPxValue,
        },
        'padding-bottom': {
          value:
            typeof bottomSpacing !== 'undefined'
              ? bottomSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.unit,
          transform: getPxValue,
        },
        'padding-left': {
          value:
            typeof leftSpacing !== 'undefined'
              ? leftSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.unit * 3,
          transform: getPxValue,
        },
        'padding-right': {
          value:
            typeof rightSpacing !== 'undefined'
              ? rightSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.unit * 3,
          transform: getPxValue,
        },
      })};
  }
`;

Root.displayName = 'ListSubheader';

const ListSubheader: React.FC<RootProps> = Root;

export { ListSubheader, RootProps as ListSubheaderProps };
