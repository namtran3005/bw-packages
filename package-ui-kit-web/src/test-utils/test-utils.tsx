import {
  render as rtlRender,
  waitFor,
  RenderOptions as RTLRenderOptions,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { WebAppThemeProvider } from '../theme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

type RenderOptions = RTLRenderOptions;

function renderWithTestHarness(
  ui: React.ReactElement,
  renderOptions: RenderOptions = {}
) {
  function Wrapper({ children }: any) {
    return (
      <MemoryRouter>
        <WebAppThemeProvider>{children}</WebAppThemeProvider>
      </MemoryRouter>
    );
  }

  return rtlRender(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// Use if you're in a hurry (better than nothing)
// be sure to verify no memory leak from implicit
// waitFor by adding these flags to jest: --runInBand --detectOpenHandles
async function renderAct(ui: React.ReactElement, options: RenderOptions = {}) {
  const view = renderWithTestHarness(ui, options);
  return await waitFor(() => view, {
    timeout: 5000,
    onTimeout: (e) => {
      // eslint-disable-next-line no-console
      console.error(e);
      return e;
    },
  });
}

export * from '@testing-library/react';

export { renderWithTestHarness, renderAct, userEvent };
