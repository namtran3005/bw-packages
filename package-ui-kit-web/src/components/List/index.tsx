import React from 'react';

import { List as ListComponent, ListProps } from './List';
import {
  ListItem as ListItemComponent,
  ListItemProps,
  ListItemIcon as ListItemIconComponent,
  ListItemIconProps,
} from './ListItem';
import {
  ListSubheader as ListSubeHeaderComponent,
  ListSubheaderProps,
} from './ListSubheader';

function List(props: ListProps) {
  return <ListComponent {...props} />;
}

function ListItem(props: ListItemProps) {
  return <ListItemComponent {...props} />;
}

const ListItemIcon = (props: any) => {
  return <ListItemIconComponent {...props} />;
};

function ListSubheader(props: ListSubheaderProps) {
  return <ListSubeHeaderComponent {...props} />;
}

List.prototype.Subheader = ListSubheader;
List.Subheader = ListSubheader;
ListItem.prototype.Icon = ListItemIcon;
ListItem.Icon = ListItemIcon;
List.prototype.Item = ListItem;
List.Item = ListItem;

export {
  List,
  ListProps,
  ListItem,
  ListItemProps,
  ListItemIcon,
  ListItemIconProps,
};
