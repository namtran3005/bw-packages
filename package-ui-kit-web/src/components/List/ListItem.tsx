import { Colors, ColorsType, Spacing, Transitions } from '@bitwala-cryptobank-squad/package-theme';
import { ListItem as MuiListItem } from '@material-ui/core';
import { ListItemProps as MuiListItemProps } from '@material-ui/core/ListItem';
import React, { Fragment } from 'react';
import styled from 'styled-components';

import {
  CardTransactionAutomobileIcon,
  ChevronRightIcon,
  QRCodeIcon
} from '@bitwala-cryptobank-squad/package-icons';
import { mediaStyledProperties } from '../../utils/mediaStyledProperty';
import { Badge } from '../Badge';
import { getPxValue, RootSpacingProps } from '../Content/Content.styled';
import { Divider } from '../Divider';
import { Typography } from '../Typography';
import { ListItemIcon, RootProps as ListItemIconProps } from './ListItemIcon';

interface RootProps extends Omit<MuiListItemProps, 'ref'>, RootSpacingProps {
  innerRef?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  selected?: boolean;
  variant?: string;
  description?: string;
  titleIcon?: React.ReactNode;
  icon?: React.ReactNode;
  badgeBackground?: ColorsType;
  textRightTop?: string;
  textRightBottom?: string;
  amountTitle?: string;
  amount?: string;
}

const highlightWidth = Spacing.unit / 2;
const doubleHighlightWidth = Spacing.unit + Spacing.unit / 2;

const RootComponent: React.FunctionComponent<RootProps> = ({
  selected,
  className,
  /* eslint-disable @typescript-eslint/no-unused-vars */
  title,
  description,
  spacing,
  horizontalSpacing,
  verticalSpacing,
  topSpacing,
  bottomSpacing,
  leftSpacing,
  rightSpacing,
  maxWidth,
  fullHeight,
  variant,
  children,
  titleIcon,
  icon,
  textRightTop,
  textRightBottom,
  amountTitle,
  amount,
  badgeBackground,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ...cleanProps
}) => {
  const ListItemInfo = () => {
    return (
      <Fragment>
        <Typography.Body style={{ marginBottom: Spacing.unit / 2 }}>
          {title}
        </Typography.Body>
        <Typography.Body>{description}</Typography.Body>
      </Fragment>
    );
  };

  const ListItemRecipient = () => {
    return (
      <Fragment>
        <WrapperContentRecipient>
          <WrapperTextRecipient>
            <Typography.Body style={{ marginBottom: Spacing.unit / 2 }}>
              {title}
            </Typography.Body>
            <Typography.Body>{description}</Typography.Body>
          </WrapperTextRecipient>
          <ChevronRightIcon foreground="primaryBlack" />
        </WrapperContentRecipient>
        <Divider />
      </Fragment>
    );
  };

  const ListItemWealth = () => {
    return (
      <ContainerWealth>
        <Badge background="success" size="small" />
        <WrapperTitleWithIcon>
          <WrapperTexts>
            <WrapperTitleWealth>
              <Typography.Body
                style={{
                  marginRight: Spacing.unit / 2,
                }}
              >
                {title}
              </Typography.Body>
              {titleIcon}
            </WrapperTitleWealth>

            <Typography.Body
              style={{
                display: 'flex',
                alignItems: 'center',
              }}
            >
              {textRightTop} <ChevronRightIcon size={16} />
            </Typography.Body>
          </WrapperTexts>
          <WrapperTexts>
            <Typography.Body>{description}</Typography.Body>
            <Typography.Body>{textRightBottom}</Typography.Body>
          </WrapperTexts>
        </WrapperTitleWithIcon>
        <Divider />
      </ContainerWealth>
    );
  };

  const ListItemTransaction = () => {
    return (
      <ContainerTransaction>
        <div style={{ flex: 1, display: 'flex' }}>
          <QRCodeIcon />
          <WrapperTitleWithIcon>
            <WrapperTexts>
              <Typography.Body>{title}</Typography.Body>
              <Typography.Body> {textRightTop}</Typography.Body>
            </WrapperTexts>
            <WrapperTexts>
              <Typography.Body>{description}</Typography.Body>
              <Typography.Body>{textRightBottom}</Typography.Body>
            </WrapperTexts>
            <BadgeContainer>
              <Typography.Body>{amountTitle}</Typography.Body>
              <Typography.Body>{amount}</Typography.Body>
            </BadgeContainer>
          </WrapperTitleWithIcon>
        </div>
        <Divider />
      </ContainerTransaction>
    );
  };

  const ListItemDefault = () => {
    return (
      <Fragment>
        <CardTransactionAutomobileIcon foreground="primaryBlack" />
        <WrapperTitleDefault>
          <Typography.Body style={{ marginBottom: Spacing.unit / 2 }}>
            {title}
          </Typography.Body>
          <Typography.Body>{description}</Typography.Body>
        </WrapperTitleDefault>
        <ChevronRightIcon foreground="primaryBlack" />
      </Fragment>
    );
  };

  const renderStructureByVariant = (variant: string | undefined) => {
    switch (variant) {
      case 'info':
        return <ListItemInfo />;
      case 'recipient':
        return <ListItemRecipient />;
      case 'wealth':
        return <ListItemWealth />;
      case 'transaction':
        return <ListItemTransaction />;
      case 'default':
        return <ListItemDefault />;
      default:
        break;
    }
  };
  // Ignoring for now as the types are broken with current version of TS; https://github.com/mui-org/material-ui/issues/14971
  return (
    <MuiListItem
      className={selected ? [className, 'selected'].join(' ') : className}
      disableRipple={variant === 'sidepanel' ? true : false}
      {...(cleanProps as any)} // eslint-disable-line @typescript-eslint/no-explicit-any
    >
      {children ? children : renderStructureByVariant(variant)}
    </MuiListItem>
  );
};

const WrapperContentRecipient = styled.div`
  flex: 1;
  display: flex;
  border-bottom: 1px solid ${Colors.getWithAlpha('text', 0.25)};
  padding-bottom: 16px;
  align-items: center;
`;

const WrapperTextRecipient = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
`;

const WrapperTitleDefault = styled.div`
  flex: 1;
  margin-left: 8px;
  display: flex;
  flex-direction: column;
`;

const WrapperTitleWithIcon = styled.div`
  flex: 1;
  margin-left: 8px;
  border-bottom: 1px solid ${Colors.getWithAlpha('text', 0.25)};
  padding-bottom: 16px;
`;

const WrapperTitleWealth = styled.div`
  display: flex;
  align-items: center;
`;
const WrapperTexts = styled.div`
  display: flex;
  justify-content: space-between;
  flex: 1;
  align-items: center;
  margin-bottom: 4px;
`;

const ContainerTransaction = styled.div`
  display: flex;
  flex: 1;
`;

const ContainerWealth = styled.div`
  display: flex;
  flex: 1;
  align-items: flex-start;
  > button {
    margin-top: 8px;
  }
`;

const BadgeContainer = styled.div`
  width: 262px;
  border-radius: 8px;
  color: ${Colors.get('success')};
  background: ${Colors.getWithAlpha('success', 0.08)};
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-size: 14px;
  padding: 6px 8px;
  max-height: 31px;
  margin-top: 16px;
`;

const Root = styled(RootComponent)`
  &&& {
    box-shadow: 0 0 0 transparent;
    transition: ${Transitions.duration.quick}ms ${Transitions.easing.swiftOut};
    ${({ topSpacing, bottomSpacing, leftSpacing, rightSpacing, spacing }) =>
      mediaStyledProperties({
        'padding-top': {
          value:
            typeof topSpacing !== 'undefined'
              ? topSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.base / 2,
          transform: getPxValue,
        },
        'padding-bottom': {
          value:
            typeof bottomSpacing !== 'undefined'
              ? bottomSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.base / 2,
          transform: getPxValue,
        },
        'padding-left': {
          value:
            typeof leftSpacing !== 'undefined'
              ? leftSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.base,
          transform: getPxValue,
        },
        'padding-right': {
          value:
            typeof rightSpacing !== 'undefined'
              ? rightSpacing
              : typeof spacing !== 'undefined'
              ? spacing
              : Spacing.base,
          transform: getPxValue,
        },
      })};
    &:hover {
      ${({ button, variant }) =>
        !button || variant === 'sidepanel'
          ? `box-shadow: none !important;`
          : ''};
      ${({ variant }) =>
        variant === 'sidepanel'
          ? `background-color: ${Colors.getWithAlpha(
              'primaryGrey',
              0.5
            )}!important`
          : ''};
    }
    &:before,
    &:after {
      content: '';
      position: absolute;
      left: 0;
      top: 0;
      height: 100%;
      width: 0;
      transition: width ${Transitions.duration.quick}ms
        ${Transitions.easing.swiftOut};
    }
    &:after {
      width: ${({ selected }) => (selected ? highlightWidth : 0)}px;
    }
    &:before {
      width: ${({ selected }) => (selected ? doubleHighlightWidth : 0)}px;
    }
  }
`;
export { Root, RootProps, highlightWidth, doubleHighlightWidth };
export {
  ListItem,
  ListItemIcon,
  ListItemIconProps,
  RootProps as ListItemProps,
};

const ListItem: React.FC<RootProps> = Root;


