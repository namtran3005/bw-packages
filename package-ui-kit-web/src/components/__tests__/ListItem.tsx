import * as React from 'react';

import { renderWithTestHarness } from '../../test-utils/test-utils';
import { ListItem } from '../List';

describe('ListItem', () => {
  it('should render ListItem', () => {
    const { baseElement } = renderWithTestHarness(
      <ListItem selected={false} />
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('should render selected ListItem', () => {
    const { baseElement } = renderWithTestHarness(<ListItem selected />);
    expect(baseElement).toMatchSnapshot();
  });
});
