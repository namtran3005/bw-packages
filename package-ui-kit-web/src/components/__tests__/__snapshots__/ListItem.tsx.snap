// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render ListItem 1`] = `
.c0.c0.c0 {
  box-shadow: 0 0 0 transparent;
  -webkit-transition: 150ms cubic-bezier(.62,.28,.23,.99);
  transition: 150ms cubic-bezier(.62,.28,.23,.99);
  padding-top: 12px;
  padding-bottom: 12px;
  padding-left: 24px;
  padding-right: 24px;
}

.c0.c0.c0:hover {
  box-shadow: none !important;
}

.c0.c0.c0:before,
.c0.c0.c0:after {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 0;
  -webkit-transition: width 150ms cubic-bezier(.62,.28,.23,.99);
  transition: width 150ms cubic-bezier(.62,.28,.23,.99);
}

.c0.c0.c0:after {
  width: 0px;
}

.c0.c0.c0:before {
  width: 0px;
}

<body>
  <div>
    <li
      class="MuiListItem-root c0 MuiListItem-gutters"
    />
  </div>
</body>
`;

exports[`ListItem should render selected ListItem 1`] = `
<body>
  .c0.c0.c0 {
  box-shadow: 0 0 0 transparent;
  -webkit-transition: 150ms cubic-bezier(.62,.28,.23,.99);
  transition: 150ms cubic-bezier(.62,.28,.23,.99);
  padding-top: 12px;
  padding-bottom: 12px;
  padding-left: 24px;
  padding-right: 24px;
}

.c0.c0.c0:hover {
  box-shadow: none !important;
}

.c0.c0.c0:before,
.c0.c0.c0:after {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 0;
  -webkit-transition: width 150ms cubic-bezier(.62,.28,.23,.99);
  transition: width 150ms cubic-bezier(.62,.28,.23,.99);
}

.c0.c0.c0:after {
  width: 4px;
}

.c0.c0.c0:before {
  width: 12px;
}

<div>
    <li
      class="MuiListItem-root c0 selected MuiListItem-gutters"
    />
  </div>
</body>
`;
